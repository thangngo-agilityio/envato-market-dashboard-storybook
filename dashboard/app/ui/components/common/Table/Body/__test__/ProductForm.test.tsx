import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { ProductForm } from '..';
import userEvent from '@testing-library/user-event';

describe('Product Modal', () => {
  const mockProduct = {
    _id: '123',
    name: 'Tivi',
    amount: 1111,
    stock: 11,
    imageURLs: ['image.png'],
  };

  it('match to snapshot', () => {
    const { container } = render(<ProductForm product={mockProduct} />);

    expect(container).toMatchSnapshot();
  });

  it('should call onCloseModal on cancel button click', async () => {
    const mockOnCloseModal = jest.fn();
    render(
      <ProductForm product={mockProduct} onCloseModal={mockOnCloseModal} />,
    );

    // Click the cancel button
    fireEvent.click(screen.getByText('Cancel'));

    waitFor(() => {
      // Ensure onCloseModal is called
      expect(mockOnCloseModal).toHaveBeenCalled();
    });
  });

  it('handleSubmitForm should update product, reset form, and close modal', async () => {
    const onUpdateProductMock = jest.fn();
    const onCloseModalMock = jest.fn();

    render(
      <ProductForm
        product={mockProduct}
        onUpdateProduct={onUpdateProductMock}
        onCloseModal={onCloseModalMock}
      />,
    );

    const inputField = screen.getByTestId('edit-field-name');

    await fireEvent.change(inputField, { target: { value: 'NewValue' } });

    await userEvent.click(screen.getByTestId('submit-product-form'));

    // Assertions
    waitFor(() => {
      expect(onUpdateProductMock).toHaveBeenCalled();
      expect(onCloseModalMock).toHaveBeenCalled();
    });
  });
});
