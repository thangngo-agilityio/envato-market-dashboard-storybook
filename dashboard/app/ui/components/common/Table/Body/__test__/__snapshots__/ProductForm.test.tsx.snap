// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Modal match to snapshot 1`] = `
<div>
  <form
    class="chakra-stack css-tl3ftk"
    id="update-product-form"
  >
    <div
      class="css-rjewin"
    >
      <div
        class="css-1veluxv"
      >
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <label
            class="chakra-form__label css-1ebhc9e"
            for="field-:r0:"
            id="field-:r0:-label"
          >
            Name
          </label>
          <div
            class="chakra-input__group css-169ufh8"
            data-group="true"
          >
            <input
              class="chakra-input css-12ku0zj"
              data-testid="edit-field-name"
              id="field-:r0:"
              name="name"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="css-1veluxv"
      >
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <label
            class="chakra-form__label css-1ebhc9e"
            for="field-:r1:"
            id="field-:r1:-label"
          >
            Price
          </label>
          <input
            autocomplete="off"
            class="chakra-input css-1iok8jb"
            data-testid="field-amount"
            id="field-:r1:"
            name="amount"
            placeholder="0.00"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="css-rjewin"
    >
      <div
        class="css-1veluxv"
      >
        <div
          class="chakra-form-control css-1vhmbn2"
          role="group"
        >
          <label
            class="chakra-form__label css-1ebhc9e"
            for="field-:r2:"
            id="field-:r2:-label"
          >
            Quantity
          </label>
          <input
            autocomplete="off"
            class="chakra-input css-1iok8jb"
            data-testid="field-quantity"
            id="field-:r2:"
            name="quantity"
            placeholder="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="css-dzz5xx"
      >
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <label
            class="chakra-form__label css-1ebhc9e"
            for="field-:r3:"
            id="field-:r3:-label"
          >
            Description
          </label>
          <div
            class="chakra-input__group css-169ufh8"
            data-group="true"
          >
            <input
              class="chakra-input css-1ua7efh"
              id="field-:r3:"
              name="description"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <label
        class="chakra-form__label css-1ebhc9e"
        for="field-:r4:"
        id="field-:r4:-label"
      >
        Currency
      </label>
      <div
        class="chakra-input__group css-169ufh8"
        data-group="true"
      >
        <input
          class="chakra-input css-1ua7efh"
          disabled=""
          id="field-:r4:"
          name="currency"
          type="text"
          value="$"
        />
      </div>
    </div>
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <div
        class="css-hk64jw"
      >
        <label
          class="chakra-form__label css-1ebhc9e"
          for="file"
          id="field-:r5:-label"
        >
          Gallery Thumbnail
          <div
            class="css-1tak1x5"
          >
            <div
              class="css-1ehgv16"
            />
            <div
              class="css-guqklo"
            />
          </div>
        </label>
        <div
          class="css-4q2zl7"
          role="presentation"
          tabindex="0"
        >
          <input
            class="chakra-input css-0"
            data-testid="field-image"
            id="field-:r5:"
            multiple=""
            style="display: none;"
            tabindex="-1"
            type="file"
          />
          <p
            class="chakra-text css-0"
          >
            Drag drop some files here, or click to select files
          </p>
        </div>
      </div>
    </div>
    <div
      class="css-16xzy9q"
    >
      <button
        class="chakra-button css-nz6i3s"
        data-testid="submit-product-form"
        disabled=""
        form="update-product-form"
        type="submit"
      >
        Save
      </button>
      <button
        class="chakra-button css-vxb15r"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;
